File: packages/greetings/greetings.py
import json
import random
from os import path
from typing import NamedTuple

import i18n


class Word(NamedTuple):
    locale: str
    word: str


def read_hellos_from_file(filepath: str) -> list[Word]:
    objs: list[Word] = []

    with open(filepath, "r") as f:
        objs = json.load(f)

    return [Word(**obj) for obj in objs]


def load_bundle(hellos: list[Word], hello_en: Word):
    for [locale, word] in hellos:
        i18n.add_translation(hello_en.word, word, locale=locale)


def greeting(greetings_file=path.join("data", "greetings.json")):
    hellos = read_hellos_from_file(greetings_file)
    hello_en = next(hello for hello in hellos if hello.locale == "en")

    if hello_en is None:
        raise Exception("No English hello found")

    load_bundle(hellos, hello_en)

    locale = random.choice(hellos).locale

    return i18n.t(hello_en.word, locale=locale)

File: packages/greetings/greetings.ts
import { readFileSync } from "fs";
import i18n, { Resource } from "i18next";
import { join } from "path";

class Word {
  locale: string;
  word: string;

  constructor(locale: string, word: string) {
    this.locale = locale;
    this.word = word;
  }
}

const readHellosFromFile = () => {
  const path = join(__dirname, "..", "..", "data", "greetings.json");
  const data = readFileSync(path, "utf8");

  const hellos: Word[] = [];

  for (const obj of JSON.parse(data)) {
    const hello = new Word(obj.locale, obj.word);

    hellos.push(hello);
  }

  return hellos;
};

const loadBundle = (hellos: Word[], helloEn: Word) => {
  const resources: Resource = {};

  for (const hello of hellos) {
    resources[hello.locale] = { [helloEn.word]: hello.word };
  }

  i18n.init({ resources });
};

export const greeting = () => {
  const hellos = readHellosFromFile();
  const helloEn = hellos.find((hello) => hello.locale === "en");

  if (!helloEn) {
    throw new Error("English greeting not found");
  }

  loadBundle(hellos, helloEn);

  const randomHello = hellos[Math.floor(Math.random() * hellos.length)];
  const randomLanguage = randomHello.locale;

  return i18n.getDataByLanguage(randomLanguage)[helloEn.word];
};
